anndata/dev/pytorch/_anndataloader.py

```diff
@@ -0,0 +1,156 @@
+from ..._core.anndata import AnnData
+from ..multi_files._anndataset import AnnDataSet, _ConcatViewMixin
+import numpy as np
+import warnings
+
+try:
+    import torch
+    from torch.utils.data import Sampler, Dataset, DataLoader
+except ImportError:
+    warnings.warn("Сould not load pytorch.")
+    Sampler, Dataset, DataLoader = object, object, object
+
+
+# Custom sampler to get proper batches instead of joined separate indices
+# maybe move to multi_files
+class BatchIndexSampler(Sampler):
+    def __init__(self, n_obs, batch_size, shuffle=False, drop_last=False):
+        self.n_obs = n_obs
+        self.batch_size = batch_size if batch_size < n_obs else n_obs
+        self.shuffle = shuffle
+        self.drop_last = drop_last
+
+    def __iter__(self):
+        if self.shuffle:
+            indices = np.random.permutation(self.n_obs).tolist()
+        else:
+            indices = list(range(self.n_obs))
+
+        for i in range(0, self.n_obs, self.batch_size):
+            batch = indices[i : min(i + self.batch_size, self.n_obs)]
+
+            # only happens if the last batch is smaller then batch_size
+            if len(batch) < self.batch_size and self.drop_last:
+                continue
+
+            yield batch
+
+    def __len__(self):
+        if self.drop_last:
+            length = self.n_obs // self.batch_size
+        else:
+            length = ceil(self.n_obs / self.batch_size)
+
+        return length
+
+
+# maybe replace use_cuda with explicit device option
+def default_converter(arr, use_cuda, pin_memory):
+    if isinstance(arr, torch.Tensor):
+        if use_cuda:
+            arr = arr.cuda()
+        elif pin_memory:
+            arr = arr.pin_memory()
+    elif arr.dtype.name != "category" and np.issubdtype(arr.dtype, np.number):
+        if issparse(arr):
+            arr = arr.toarray()
+        if use_cuda:
+            arr = torch.tensor(arr, device="cuda")
+        else:
+            arr = torch.tensor(arr)
+            arr = arr.pin_memory() if pin_memory else arr
+    return arr
+
+
+def _convert_on_top(convert, top_convert, attrs_keys):
+    if convert is None:
+        new_convert = top_convert
+    elif callable(convert):
+        new_convert = lambda arr: top_convert(convert(arr))
+    else:
+        new_convert = {}
+        for attr in attrs_keys:
+            if attr not in convert:
+                new_convert[attr] = top_convert
+            else:
+                if isinstance(attrs_keys, list):
+                    as_ks = None
+                else:
+                    as_ks = attrs_keys[attr]
+                new_convert[attr] = _convert_on_top(convert[attr], top_convert, as_ks)
+    return new_convert
+
+
+# AnnDataLoader has the same arguments as DataLoader, but uses BatchIndexSampler by default
+class AnnDataLoader(DataLoader):
+    def __init__(
+        self,
+        adatas,
+        batch_size=1,
+        shuffle=False,
+        use_default_converter=True,
+        use_cuda=False,
```

Comment:
pl lightning and other frameworks handle moving to\out of cuda\specific gpu automatically + if using multiple gpus tensors need to be sent separately.

Reply:
Hi, yes, i know, this is why it is set to `False` by default.